### PR TITLE
Fix for issue #380

### PR DESCRIPTION
**Pull request by VincentDeHaen**

### CODE REVIEW

File: stack/pico_device.c [hunk 1]
```diff
@@ -166,7 +166,14 @@ int pico_device_init(struct pico_device *dev, const char *name, const uint8_t *m
         return -1;
     }
 
-    pico_tree_insert(&Device_tree, dev);
+    if(pico_tree_insert(&Device_tree, dev)){
+		if(pico_err != PICO_ERR_ENOMEM){
+			pico_err = PICO_ERR_EINVAL;
+		}
+		//PICO_FREE(dev->q_in);
+		//PICO_FREE(dev->q_out);
```
Comment (frederikvs) [inline stack/pico_device.c]: Why are these commented out?

File: modules/pico_fragments.c [hunk 2]
```diff
@@ -430,7 +430,14 @@ void pico_ipv6_process_frag(struct pico_ipv6_exthdr *frag, struct pico_frame *f,
...
     {
       if ((pico_ipv6_frag_match(f, first) == 0 && (IP6_FRAG_ID(frag) == ipv6_cur_frag_id))) {
-        pico_tree_insert(&ipv6_fragments, pico_frame_copy(f));
+        if(pico_tree_insert(&ipv6_fragments, pico_frame_copy(f)) ){
+			frag_dbg("Could not insert picoframe in tree\n");
+			if(pico_err != PICO_ERR_ENOMEM){
+				frag_dbg("Key is already in tree\n");
+				pico_err = PICO_ERR_EINVAL;
+			}
+			return;
```
Comment (frederikvs) [inline modules/pico_fragments.c]: Shouldn't you delete the result of the `pico_frame_copy` here? (same goes for other occurrences in this file)
And actually, the frame copy could go wrong as well (also due to memory issues), so you might want to check for that as well, before trying to insert.

File: modules/pico_mcast.c [hunk 3]
```diff
@@ -73,7 +73,11 @@ static int pico_mcast_src_filtering_inc_inc(struct mcast_filter_parameters*mcast
         if (mcast->p->MCASTFilter) {
             pico_tree_foreach(index, mcast->p->MCASTFilter) /* B */
             {
-                pico_tree_insert(mcast->allow, index->keyValue);
+                if(pico_tree_insert(mcast->allow, index->keyValue) ){
+               		if(pico_err != PICO_ERR_ENOMEM){
+               			pico_err = PICO_ERR_EINVAL;
+               		}
+               	}
```
Comment (frederikvs) [inline modules/pico_mcast.c]: You don't signal the failure e.g. by a `return` or similar - is this by design?

File: modules/pico_dns_sd.c [hunk 4]
```diff
@@ -426,10 +426,22 @@ pico_dns_sd_register_service( const char *name,
...
 
     if (txt_record) {
-        pico_tree_insert(&rtree, txt_record);
+        if(pico_tree_insert(&rtree, txt_record)){
+        	if(pico_err != PICO_ERR_ENOMEM){
+        		pico_err = PICO_ERR_EINVAL;
+            }
+        	pico_tree_destroy(&rtree, NULL);
+            return -1;
+        }
```
Comment (frederikvs) [inline modules/pico_dns_sd.c]: Don't you need to delete the `txt_record` as well? It was created a little higher, but not inserted. Also, you might want to use the `PICO_MDNS_RTREE_DESTROY` macro here, as is done later on.
(same comments apply to the code below, regarding `srv_record`)

File: modules/pico_mdns.c [hunk 5]
```diff
@@ -592,8 +592,13 @@ pico_mdns_rtree_find_name_type( pico_mdns_rtree *tree,
...
 
-            if (record)
-                pico_tree_insert(&hits, record);
+            if (record){
+            	if(pico_tree_insert(&hits, record) ){
+            		if(pico_err != PICO_ERR_ENOMEM){
+            			pico_err = PICO_ERR_EINVAL;
+            		}
+            	}
+            }
```
Comment (frederikvs) [inline modules/pico_mdns.c]: No freeing of the record?
No `return` when it fails?

File: modules/pico_mdns.c [hunk 6]
```diff
@@ -762,7 +767,13 @@ pico_mdns_generate_new_records( pico_mdns_rtree *conflict_records,
...
 
             /* Add the record to the new tree */
-            pico_tree_insert(&new_records, new_record);
+            if(pico_tree_insert(&new_records, new_record) ){
+            	mdns_dbg("Could not add new non-conflicting record to the tree!\n");
+				if(pico_err != PICO_ERR_ENOMEM){
+					pico_err = PICO_ERR_EINVAL;
+				}
+				return new_records;
+			}
```
Comment (frederikvs) [inline modules/pico_mdns.c]: Doesn't the `new_record` need to be freed too, somehow? Maybe @jelledevleeschouwer can shed some light on this?

File: modules/pico_mdns.c [hunk 7]
```diff
@@ -807,8 +818,14 @@ pico_mdns_record_resolve_conflict( struct pico_mdns_record *record,
...
-    if (copy)
-        pico_tree_insert(&new_records, copy);
+    if (copy){
+    	if(pico_tree_insert(&new_records, copy) ){
+			if(pico_err != PICO_ERR_ENOMEM){
+				pico_err = PICO_ERR_EINVAL;
+			}
+			return -1;
+		}
+    }
```
Comment (frederikvs) [inline modules/pico_mdns.c]: I think you need to delete the `copy` here as well.

File: modules/pico_mdns.c [hunk 8]
```diff
@@ -2356,7 +2404,12 @@ pico_mdns_additionals_add_nsec( pico_mdns_rtree *artree,
 
     /* If there is none present generate one for given name */
     if ((nsec = pico_mdns_gen_nsec_record(name)))
-        pico_tree_insert(artree, nsec);
+    	if(pico_tree_insert(artree, nsec) ){
+			if(pico_err != PICO_ERR_ENOMEM){
+				pico_err = PICO_ERR_EINVAL;
+			}
+			return -1;
+		}
```
Comment (frederikvs) [inline modules/pico_mdns.c]: I think `nsec` is allocated in `pico_mdns_gen_nsec_record` (correct me if I'm wrong), so you'll probably need to free it as well.

File: modules/pico_mdns.c [hunk 9]
```diff
@@ -2927,7 +2980,14 @@ pico_mdns_getrecord_generic( const char *url, uint16_t type,
...
 
-    pico_tree_insert(&qtree, q);
+    if(pico_tree_insert(&qtree, q) ){
+    	mdns_dbg("inserting query into tree failed!\n");
+		if(pico_err != PICO_ERR_ENOMEM){
+			pico_err = PICO_ERR_EINVAL;
+		}
+		return -1;
+	}
+
```
Comment (frederikvs) [inline modules/pico_mdns.c]: need to free `q`?

File: modules/pico_mdns.c [hunk 10]
```diff
@@ -2940,7 +3000,13 @@ pico_mdns_getrecord_generic( const char *url, uint16_t type,
...
 
     /* Add cookie to Cookies to be able to find it afterwards */
-    pico_tree_insert(&Cookies, cookie);
+    if(pico_tree_insert(&Cookies, cookie) ){
+		mdns_dbg("inserting cookie into tree failed!\n");
+		if(pico_err != PICO_ERR_ENOMEM){
+			pico_err = PICO_ERR_EINVAL;
+		}
+		return -1;
+	}
```
Comment (frederikvs) [inline modules/pico_mdns.c]: need to free the `cookie`?

**Comment by frederikvs** [comment]
Thanks for your work!

I've added some questions/thoughts to the diffs.

One global thought : Wouldn't it be better to check the actual return value of `pico_tree_insert` (`&LEAF` if  `ENOMEM` vs. the node that was already in there), than to check the value of `pico_err`? If some previous function ran out of memory, but the `pico_err` was never cleared, you might take the wrong decision.

File: stack/pico_device.c (hunk 1 above)
Comment (VincentDeHaen) [inline stack/pico_device.c]: Sorry my bad! Fixed this one!

**Comment by frederikvs** [comment]
Finally found the time to review some of these changes. Sorry it took so long.

This PR looks good to me, bonus points for `pico_protocol_init`, keeping the code readable and nice.

Minor remark : in some places, tabs were used for indentation, in others spaces. I've decided to accept the PR regardless, since pretty soon we'll run `make style` on it again, but next time you might want to check your editor settings. Spaces are correct :-)

Thanks for the hard work!

**Pull request merged**